### PR TITLE
3453 download lup docs

### PR DESCRIPTION
**Pull request by allthesignals**
### Summary

Query for and display LUP's (a disposition record's) related documents, and enable download, according to certain filter criteria: 
image

#### Tasks/Bug Numbers
 - Fixes [AB#3453](https://dev.azure.com/NYCPlanning/cc280b0d-40a0-4689-b852-2e6247f1af50/_workitems/edit/3453)

 - Related [AB#3388](https://dev.azure.com/NYCPlanning/cc280b0d-40a0-4689-b852-2e6247f1af50/_workitems/edit/3388)

### Technical Explanation
Nothing super idiosyncratic here. This follows the same pattern for doc download/display as packages and artifacts.

### CODE REVIEW

File: server/src/document/document.controller.ts [hunk 1]
```diff
@@ -1,4 +1,5 @@
-import { Controller,
+import {
```
Comment (godfreyyeung) [inline server/src/document/document.controller.ts:1]: Are changes in this file all linting?

File: server/src/document/document.service.ts [hunk 2]
```diff
@@ -118,59 +133,115 @@ export class DocumentService {
...
+    try {
+      const {
+        records: [firstDisposition]
+      } = await this.crmService.get(
+        "dcp_communityboarddispositions",
+        `
+        $select=dcp_communityboarddispositionid
+        &$filter=dcp_communityboarddispositionid eq ${recordId}
+          and (
+            dcp_visibility eq ${DISPOSITION_VISIBILITY.GENERAL_PUBLIC}
```
Comment (godfreyyeung) [inline server/src/document/document.service.ts:227]: cool!

Comment (godfreyyeung) [reply server/src/document/document.service.ts:227]: Thanks for all this research!

File: server/src/project/project.service.ts [hunk 3]
```diff
@@ -507,6 +509,40 @@ export class ProjectService {
...
+    let projectDispositions =
+      transformedProject[
+        "dcp_dcp_project_dcp_communityboarddisposition_project"
+      ];
+
+    try {
+      projectDispositions = await Promise.all(
+        projectDispositions.map(async disposition => {
+          if (
+            [717170004, 717170003].includes(disposition.dcp_visibility) && // LUP or General Public
```
Comment (godfreyyeung) [inline server/src/project/project.service.ts:521]: tiny tiny nit: can we use the consts here?

File: server/src/project/project.service.ts [hunk 4]
```diff
@@ -620,7 +656,7 @@ export class ProjectService {
 
       dispositions: {
         ref: "dcp_communityboarddispositionid",
-        attributes: DISPOSITION_KEYS
```
Comment (godfreyyeung) [inline server/src/project/project.service.ts:623]: why not follow same pattern as PACKAGE and ARTIFACTs, and place this key inside their constant lists?

File: server/src/project/project.service.ts (hunk 3 above)
Comment (allthesignals) [reply server/src/project/project.service.ts:521]: yep!

File: server/src/document/document.controller.ts (hunk 1 above)
Comment (allthesignals) [reply server/src/document/document.controller.ts:1]: uh oh. yep except for the bottom of the file.

File: server/package.json [hunk 5]
```diff
@@ -83,6 +83,7 @@
   },
   "devDependencies": {
     "@nestjs/cli": "^7.6.0",
+    "@nestjs/schematics": "^7.3.1",
```
[the remark this answers is not in the document]
Comment (allthesignals) [reply server/package.json:86]: yes

File: server/src/disposition/disposition.service.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+  /**
+   * @param      {string}  relativeUrl    a string representing the Sharepoint URL relative to the entity folder.
+   *                                      In essence it is the Sharepoint folder name. e.g. P2015K0223_DOB Letter_3
+   * @param      {Object}  dcp_name   The disposition dcp_name property.
+   *                                  Used for error logging purposes.
+   *
+   * @return     {Object[]}     Array of 0 or more custom Document objects
+   */
+  async getdispositionSharepointDocuments(relativeUrl, dcp_name) {
```
[the remark this answers is not in the document]
Comment (allthesignals) [reply server/src/disposition/disposition.service.ts:16]: oops ty

File: server/src/disposition/disposition.service.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    if (relativeUrl) {
+      try {
+        const documents = await this.sharepointService.getSharepointFolderFiles(
+          `dcp_dispositions/${relativeUrl}`,
```
[the remark this answers is not in the document]
Comment (allthesignals) [reply server/src/disposition/disposition.service.ts:20]: @godfreyyeung if it doesn't work (there are download links that don't download) I need to know about that

File: server/src/project/project.service.ts (hunk 4 above)
Comment (allthesignals) [reply server/src/project/project.service.ts:623]: Yeah, I think those other lists are probably fine but the problem is that they reference non-CRM attributes (like "documents") which we later learned in applicant portal was a problem for re-using them when building the CRM queries ($select=documents would give an error).

**Pull request merged**